apiary-metastore-metrics/src/main/java/com/expedia/apiary/extensions/metastore/metrics/CodahaleMetrics.java

```diff
@@ -0,0 +1,439 @@
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Splitter;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.Lists;
+
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.permission.FsPermission;
+import org.apache.hadoop.hive.common.metrics.common.MetricsConstant;
+import org.apache.hadoop.hive.common.metrics.common.MetricsScope;
+import org.apache.hadoop.hive.common.metrics.common.MetricsVariable;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.metrics2.lib.DefaultMetricsSystem;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.lang.management.ManagementFactory;
+import java.net.URI;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TimerTask;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.apache.hadoop.hive.common.metrics.metrics2.MetricsReporting;
+import org.apache.hadoop.hive.common.metrics.metrics2.MetricVariableRatioGauge;
+
+import com.amazonaws.services.cloudwatch.AmazonCloudWatch;
+import com.amazonaws.services.cloudwatch.AmazonCloudWatchClientBuilder;
+import com.amazonaws.services.cloudwatch.model.MetricDatum;
+import com.amazonaws.services.cloudwatch.model.PutMetricDataRequest;
+import com.amazonaws.services.cloudwatch.model.PutMetricDataResult;
+import com.amazonaws.services.cloudwatch.model.StandardUnit;
+import com.amazonaws.services.cloudwatch.model.Dimension;
+
+import java.util.SortedMap; 
+
+
+/**
+ * Codahale-backed Metrics implementation.
+ */
+public class CodahaleMetrics implements org.apache.hadoop.hive.common.metrics.common.Metrics {
+
+  public static final Logger LOGGER = LoggerFactory.getLogger(CodahaleMetrics.class);
+
+  public final MetricRegistry metricRegistry = new MetricRegistry();
+  private final Lock timersLock = new ReentrantLock();
+  private final Lock countersLock = new ReentrantLock();
+  private final Lock gaugesLock = new ReentrantLock();
+  private final Lock metersLock = new ReentrantLock();
+
+  private LoadingCache<String, Timer> timers;
+  private LoadingCache<String, Counter> counters;
+  private LoadingCache<String, Meter> meters;
+  private ConcurrentHashMap<String, Gauge> gauges;
+
+  private HiveConf conf;
+  private final Set<Closeable> reporters = new HashSet<Closeable>();
+
+  private final ThreadLocal<HashMap<String, CodahaleMetricsScope>> threadLocalScopes
```

Comment:
Can HashMap not be changed to Map so it's not so tightly bound to the implementation?

Reply:
OK, other people's code, let's leave these.